Sources/Workspace/Workspace.swift

```diff
@@ -157,56 +166,96 @@ extension WorkspaceDelegate {
     }
 }
 
+private class WorkspaceManifestLoaderDelegate: ManifestLoader.Delegate {
+    private weak var workspaceDelegate: Workspace.Delegate?
+
+    init(workspaceDelegate: Workspace.Delegate) {
+        self.workspaceDelegate = workspaceDelegate
+    }
+
+    func willLoad(packageIdentity: PackageIdentity, packageLocation: String, manifestPath: AbsolutePath) {
+        // handled by workspace directly
+    }
+
+    func didLoad(packageIdentity: PackageIdentity, packageLocation: String, manifestPath: AbsolutePath, duration: DispatchTimeInterval) {
+        // handled by workspace directly
+    }
+
+    func willParse(packageIdentity: PackageIdentity, packageLocation: String) {
+        // noop
+    }
+
+    func didParse(packageIdentity: PackageIdentity, packageLocation: String, duration: DispatchTimeInterval) {
+        // noop
+    }
+
+    func willCompile(packageIdentity: PackageIdentity, packageLocation: String, manifestPath: AbsolutePath) {
+        self.workspaceDelegate?.willCompileManifest(packageIdentity: packageIdentity, packageLocation: packageLocation)
+    }
+    
+    func didCompile(packageIdentity: PackageIdentity, packageLocation: String, manifestPath: AbsolutePath, duration: DispatchTimeInterval) {
+        self.workspaceDelegate?.didCompileManifest(packageIdentity: packageIdentity, packageLocation: packageLocation, duration: duration)
+    }
+
+    func willEvaluate(packageIdentity: PackageIdentity, packageLocation: String, manifestPath: AbsolutePath) {
+        self.workspaceDelegate?.willCompileManifest(packageIdentity: packageIdentity, packageLocation: packageLocation)
+    }
+
+    func didEvaluate(packageIdentity: PackageIdentity, packageLocation: String, manifestPath: AbsolutePath, duration: DispatchTimeInterval) {
+        self.workspaceDelegate?.didEvaluateManifest(packageIdentity: packageIdentity, packageLocation: packageLocation, duration: duration)
+    }
+}
+
 private class WorkspaceRepositoryManagerDelegate: RepositoryManager.Delegate {
-    private unowned let workspaceDelegate: Workspace.Delegate
+    private weak var workspaceDelegate: Workspace.Delegate?
```

Comment:
not sure I understand this change, the delegate is not optional

Reply:
`weak` vars have to be optional, don't they?

Reply:
Yes, what I mean is that the initializer requires passing a delegate, so there's no need for it to be optional. I believe that's why this is currently `unowned`. Changing it to `weak` seems odd, because there's no need for it to be weak

Reply:
https://github.com/apple/swift-package-manager/pull/6863#discussion_r1306153313